### PR TITLE
Automation/Ansible/Repositories page model

### PR DESCRIPTION
**Pull request by quarckster**
Initial implementation of Automation/Ansible/Repositories page model.
{{pytest: -v -k 'test_embedded_ansible_repository_crud' --long-running}}

### CODE REVIEW

File: cfme/ansible/repositories.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+@navigator.register(Repository)
+class Details(CFMENavigateStep):
+    VIEW = RepositoryDetailsView
+    prerequisite = NavigateToAttribute("appliance.server", "AnsibleRepositories")
+
+    def step(self):
+        for row in self.prerequisite_view.repositories:
+            if row["Name"].text == self.obj.name:
+                row["Name"].click()
```
Comment (mfalesni) [inline cfme/ansible/repositories.py]: I guess this for deserves else+raise

File: cfme/ansible/repositories.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+from widgetastic_manageiq import Table
+from widgetastic_patternfly import Dropdown, Button, Input, FlashMessages
+from utils.appliance import Navigatable
+from utils.appliance.implementations.ui import navigator, navigate_to, CFMENavigateStep
+from utils.wait import wait_for
+
+
+class RepositoryBaseView(BaseLoggedInPage):
+    flash = FlashMessages('.//div[starts-with(@class, "flash_text_div") or @id="flash_text_div"]')
+    title = Text(locator=".//h1")
```
Comment (izapolsk) [inline cfme/ansible/repositories.py]: REQUIRED: this is too broad locator. I'd advice you to use locator like 
`title = Text('//div[@id="main-content"]//h1')`

File: cfme/tests/ansible/test_embedded_ansible_basic.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    pytest.mark.long_running,
+    pytest.mark.meta(server_roles=["+embedded_ansible"]),
+    pytest.mark.uncollectif(current_version() < "5.8"),
+    pytest.mark.ignore_stream("upstream"),
+    test_requirements.ansible
+]
+
+
+@pytest.fixture(scope="module")
+def embedded_ansible_role_enabled(appliance):
```
Comment (izapolsk) [inline cfme/tests/ansible/test_embedded_ansible_basic.py]: Please consider moving this method to appliance obj.

File: cfme/ansible/repositories.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        return (
+            self.in_ansible_repositories and
+            self.title.text == 'Edit Repository "{}"'.format(self.context["object"].name)
+        )
+
+
+class RepositoryCollection(Navigatable):
+    """Collection object for the :py:class:`Repository`."""
+
+    def create(self, name, url, description=None, scm_credentials=None, scm_branch=None,
```
Comment (izapolsk) [inline cfme/ansible/repositories.py]: shouldn't here be implemented such methods as delete, all, etc ?

File: cfme/ansible/repositories.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
+# -*- coding: utf-8 -*-
+"""Page model for Automation/Anisble/Repositories"""
```
Comment (mfalesni) [inline cfme/ansible/repositories.py]: Eher si a ytpo :)

File: cfme/ansible/repositories.py (hunk 4 above)
Comment (mfalesni) [inline cfme/ansible/repositories.py]: @izapolsk eventually yes, but we can start without it.

Comment (mfalesni) [inline cfme/ansible/repositories.py]: @quarckster Though I think we could make all parameters optional to enable us testing that it is indeed required by the UI :)

File: cfme/ansible/repositories.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+            'Add of Repository "{}" was successfully initialized.'.format(name))
+
+        def _wait_until_appeared():
+            for row in repo_list_page.repositories:
+                if row["Name"].text == name:
+                    return True
+            else:
+                return False
+
+        wait_for(_wait_until_appeared, delay=10, fail_func=repo_list_page.browser.selenium.refresh)
```
Comment (mfalesni) [inline cfme/ansible/repositories.py]: This sounds like a RFE for devs for a reload button for the screen :)

File: cfme/tests/ansible/test_embedded_ansible_basic.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+from utils.update import update
+from utils.version import current_version
+from utils.appliance.implementations.ui import navigate_to
+from cfme.ansible.repositories import RepositoryCollection
+from cfme import test_requirements
+
+pytestmark = [
+    pytest.mark.long_running,
+    pytest.mark.meta(server_roles=["+embedded_ansible"]),
+    pytest.mark.uncollectif(current_version() < "5.8"),
```
Comment (mfalesni) [inline cfme/tests/ansible/test_embedded_ansible_basic.py]: This is import-time appliance access. Lambda please :)

File: cfme/tests/ansible/test_embedded_ansible_basic.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        "https://github.com/quarckster/ansible_playbooks",
+        description=fauxfactory.gen_alpha()
+    )
+    updated_description = "edited_{}".format(fauxfactory.gen_alpha())
+    with update(repository):
+        repository.description = updated_description
+    view = navigate_to(repository, "Edit")
+    wait_for(lambda: view.description.value != "", delay=1, timeout=5)
+    assert view.description.value == updated_description
+    repository.delete()
```
Comment (mfalesni) [inline cfme/tests/ansible/test_embedded_ansible_basic.py]: I would maybe say to put a finalizer in this test that deletes the repo anyway if it fails before the actual deletion?

Comment (mfalesni) [inline cfme/tests/ansible/test_embedded_ansible_basic.py]: like:

```python
repository = ...

@request.addfinalizer
def _delete_if_exists():
    if repository.exists:
        repository.delete()
```

**Comment by dajoRH** [comment]
Lint report for commit 5b10f81d65c7c16d236bc1afa0f0c61052e4eaee:

`cfme/ansible/repositories.py`:
- :bangbang: Line 120:17: **E126** *continuation line over-indented for hanging indent*
- :bangbang: Line 129:13: **E121** *continuation line under-indented for hanging indent*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 2210461acbfbb63a1073e0e9342461b1e19bc2ac:

`cfme/ansible/repositories.py`:
- :heavy_exclamation_mark: Line 93:1: **W293** *blank line contains whitespace*
- :heavy_exclamation_mark: Line 100:32: **W291** *trailing whitespace*
- :heavy_exclamation_mark: Line 103:1: **W293** *blank line contains whitespace*
- :heavy_exclamation_mark: Line 147:1: **W293** *blank line contains whitespace*
- :heavy_exclamation_mark: Line 169:1: **W293** *blank line contains whitespace*
- :heavy_exclamation_mark: Line 173:1: **W293** *blank line contains whitespace*
- :heavy_exclamation_mark: Line 225:1: **W293** *blank line contains whitespace*

Hmmm
:frowning: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

**Comment by mfalesni** [comment]
QC things reviewed, not much we can do now about it.

**Pull request merged**